src/lib.rs

```diff
@@ -80,7 +82,10 @@ impl Connection {
             let header = PacketHeader::parse(&complete_packet[0..HEADER_SIZE])?;
 
             // Read the packet payload.
-            let payload_length = header.get_payload_length() as usize;
+            let mut payload_length = header.get_payload_length() as usize;
+            if header.get_mt() == MessageType::Data {
+                payload_length = usize::from_le_bytes(complete_packet[2..3].try_into().unwrap());
+            }
```

Comment:
Here we can do pattern matching to avoid mutable variable:
```suggestion
            let payload_length = match header.get_mt() {
                MessageType::Data => usize::from_le_bytes(complete_packet[2..3].try_into().unwrap()),
                _ =>  header.get_payload_length() as usize
            };
```